Adds working tests for SoftTakeover and corrections to Time class.

Pegasus-RPG opened the pull request:
Also replaced the softTakeover::ignore() algorithm with a complete one that passes all tests. (The old one failed 8 of them.)

Pegasus-RPG committed:
WIP: Changed the softtakeover algorithm but having problems getting the tests to work correctly.

Pegasus-RPG committed:
Clarified units in Time class, allowing me to fix problems with softTakeover tests. My new code now passes them all! (Old code failed 8.)

ywwg commented on src/controllers/softtakeover.cpp:
```diff
@@ -102,32 +102,59 @@ bool SoftTakeover::ignore(ControlObject* control, double newParameter) {
+        if (fabs(difference) > m_dThreshold) {
+            // New parameter is far away from current
+            const double prevDiff = currentParameter - m_prevParameter;
+            if ((prevDiff > 0 && difference < 0) ||
```
I still don't like this as a test for "has it changed sides".  Can you try with:

if (prevDiff * difference < 0)

and see if tests pass?

Pegasus-RPG replied:
Sure. This way just made it clear what's going on. Tests pass.

ywwg replied:
(for the record)  yeah I just didn't like that the zero cases weren't handled. changing >0 to >=0 would also be fine

ywwg commented on src/controllers/softtakeover.cpp:
```diff
@@ -102,32 +102,59 @@ bool SoftTakeover::ignore(ControlObject* control, double newParameter) {
+            if ((prevDiff > 0 && difference < 0) ||
+                (prevDiff < 0 && difference > 0)) {
+                // Opposite sides
+                if ((currentTime - m_time) > SUBSEQUENT_VALUE_OVERRIDE_TIME_MILLIS) {
```
not sure if this is a new constant, but I think we're using constants like kSubsequentValueOverrideTimeMillis  (ALL_CAPS are for macros)

Pegasus-RPG replied:
It's not, but do you want me to make that change too?

ywwg replied:
up to you

ywwg commented on src/controllers/softtakeover.cpp:
```diff
@@ -102,32 +102,59 @@ bool SoftTakeover::ignore(ControlObject* control, double newParameter) {
+                    //    << "on same side of" << m_prevParameter 
+                    //    << "which is also far from current";
+                } else if ((currentTime - m_time) > SUBSEQUENT_VALUE_OVERRIDE_TIME_MILLIS) {
+                    // New parameter arrived awhile after the last one that took effect
```
awhile -> "long enough"

ywwg commented on src/controllers/softtakeover.cpp:
```diff
@@ -102,32 +102,59 @@ bool SoftTakeover::ignore(ControlObject* control, double newParameter) {
+                (prevDiff < 0 && difference > 0)) {
+                // Opposite sides
+                if ((currentTime - m_time) > SUBSEQUENT_VALUE_OVERRIDE_TIME_MILLIS) {
+                    // New parameter arrived awhile after the last one that took effect
```
awhile -> long enough

ywwg commented on src/controllers/softtakeover.cpp:
```diff
@@ -137,5 +164,5 @@ bool SoftTakeover::ignore(ControlObject* control, double newParameter) {
 
 void SoftTakeover::ignoreNext() {
-    m_time = 0;
+    m_time = -1;
```
document in the header file that -1 is a magic number meaning "uninitialized value"

ywwg commented on src/test/softtakeover_test.cpp:
```diff
@@ -550,9 +546,9 @@ TEST_F(SoftTakeoverTest, PrevFarMore_NewFarMore_Late) {
-    Time::setTestElapsedTime(SoftTakeover::TestAccess::getTimeThreshold()*2);
-    EXPECT_TRUE(st_control.ignore(co.data(), 100));
+    EXPECT_TRUE(st_control.ignore(co.data(), co->getParameterForValue(120)));
+    Time::setTestElapsedMsecs(SoftTakeover::TestAccess::getTimeThreshold()*2);
```
optional -- style guide says to put spaces around * in multiplication: SoftTakeover::TestAccess::getTimeThreshold() * 2

Pegasus-RPG committed:
- Corrected time scales in Time class (the timer is in nanoseconds!)
- Addressed PR notes
- Added tests for quick whipping of controls at the edges of the ControlObjects (when the previous value matches the current)

Pegasus-RPG committed:
Comment grammar fix

daschuer commented on src/controllers/softtakeover.cpp:
```diff
@@ -102,32 +102,74 @@ bool SoftTakeover::ignore(ControlObject* control, double newParameter) {
+        if (m_prevParameter != -1 && fabs(difference) > m_dThreshold) {
+            // New parameter is far away from current
+            const double prevDiff = currentParameter - m_prevParameter;
+            if (prevDiff * difference <= 0) {
```
This condition becomes true if currentParameter == m_prevParameter, which is IMHO wrong.

This means we ignore a value from a synced (snapped in) hardware control that it  3 midi units after 50 ms have passed.

We should never ignore values from controls once we are in sync.

daschuer replied:
Ah, nevermind. the condition is catched below. 

You can use 
`if (prevDiff * difference < 0) {`
here

daschuer replied:
ahh, no ... 

How about check for if (prevDiff == 0) first. This would make it easier to track the logic.

Pegasus-RPG replied:
I thought about that. I need to figure out which is more likely so more conditional checks can be skipped.

daschuer commented on src/controllers/softtakeover.cpp:
```diff
@@ -102,32 +102,74 @@ bool SoftTakeover::ignore(ControlObject* control, double newParameter) {
-        m_time = 1;
-//         qDebug() << "SoftTakeover::ignore: ignoring the first value"
-//                  << newParameter;
-    } else if ((currentTime - m_time) > SUBSEQUENT_VALUE_OVERRIDE_TIME_MILLIS) {
```
Moving this to the places below, will ignore some more values as before. 
* if same side changes above thresholds

I have never understood, why we need the timer at all. What is the use case for this?

Pegasus-RPG replied:
Slow-refresh controllers. If a control is moved quickly, slow controllers will send values that are far apart, causing Mixxx to lose sync. If we get two values within a short amount of time, regardless of what they are, we use them (except in just a couple cases.)

daschuer replied:
Maybe this is a relict from an earlier version. Now, once we are ins snyc big changes are not ignored even if they arrive after the 50 ms delay has passed.   

It looks like the timer has an effect if we have concurrent control commands. Like automatic pitch changes during you change the pitch from the controller as well. 
If the automatic change happens during a controller slider transaction it is ignored. Without this timer, the controller slider will falls out of sync. If the automatic change happens before or after the manual control, the hardware slider falls out of sync anyway. Is it that what we want? 

If yes, we should add a comment about that.

daschuer commented on src/controllers/softtakeover.cpp:
```diff
@@ -102,32 +102,74 @@ bool SoftTakeover::ignore(ControlObject* control, double newParameter) {
+                }
+            } else {
+                // Same side
+                if (fabs(prevDiff) > m_dThreshold) {
```
this condition is always true

Pegasus-RPG replied:
How?

daschuer commented:
We now ignore different side changes. This is new,and likely prevents that a control is reliably able to get back in sync. 
We have to be careful to change the logic how softtakover works. Since the old code was tested as good during the beta phase. I might be that the 8 failing test are false positives.

daschuer commented on src/controllers/softtakeover.cpp:
```diff
@@ -102,32 +102,74 @@ bool SoftTakeover::ignore(ControlObject* control, double newParameter) {
-//                  << newParameter;
-    } else if ((currentTime - m_time) > SUBSEQUENT_VALUE_OVERRIDE_TIME_MILLIS) {
-        // don't ignore value if a previous one was not ignored in time
+        m_time = -2;    // Tests like to use 0
```
-2 is not enough for the fast tests use -kSubsequentValueOverrideTimeMsecs to be sure.

Pegasus-RPG replied:
Tests use positive values, they all start at 10.

daschuer commented:
Old algorithm: 

[  FAILED  ] SoftTakeoverTest.SuperFastNotSame
   -> Here we cross the current value so it should be not ignored.  Test is should be fixed. 
[  FAILED  ] SoftTakeoverTest.PrevNearLess_NewFarLess_Late
   -> Here we are in the snap in zone, but do not cross. Fixing this fixes the small back jump when snapping in. 
[  FAILED  ] SoftTakeoverTest.PrevNearLess_NewFarMore_Late
   -> Here we are in the snap in zone, and cross the current value. It should snap in. Test is should be fixed. 
[  FAILED  ] SoftTakeoverTest.PrevNearMore_NewFarLess_Late
   -> Here we cross the current value so it should be not ignored.  Test is should be fixed. 
[  FAILED  ] SoftTakeoverTest.PrevNearMore_NewFarMore_Late
   -> Here we are in the snap in zone, but do not cross. Fixing this fixes the small back jump when snapping in. 
[  FAILED  ] SoftTakeoverTest.PrevFarLess_NewFarLess_Soon
   -> Fails because of the invalid timer value is to big for the test 
[  FAILED  ] SoftTakeoverTest.PrevFarLess_NewFarMore_Late
   -> Here we cross the current value so it should be not ignored.  Test is should be fixed. 
[  FAILED  ] SoftTakeoverTest.PrevFarMore_NewFarMore_Soon
   -> Fails because of the invalid timer value is to big for the test 
[  FAILED  ] SoftTakeoverTest.PrevFarMore_NewFarLess_Late
   -> Here we cross the current value so it should be not ignored.  Test is should be fixed. 

Except of the small snap jump, the old algorithm was working well. 
I think we need the snap in "area" to match the gui raster with the midi raster in weird conditions. 
However in my tests it works nicely in with a threshold of 0. 
Maybe we can reduce the threshold a bit 

If we where not in a rc1 state, I would remove the threshold and the timer, but since we have no real bug for this we should not touch the alogithm.

ywwg commented:
> we should not touch the algorithm.

That sounds like the right thing to do, especially as these changes have sort of gotten out of hand -- if we're pretty sure the old algorithm was working ok, we should construct tests such that they pass and focus just on fixing the scripting bug.  If we want to change the behavior we can do that in a PR for 2.1.

Pegasus-RPG commented:
I agree as well, since I'm not sure the tests are totally correct. I would appreciate a review of my spreadsheet in PR #806. Closing this now and will re-open against master/2.1 when #806 is merged.
